### PR TITLE
FFT::Radix2 uses bit-reversed permutation algorithm

### PR DESCRIPTION
**Pull request by lamphamsy**
Implement FFT Radix2 using bit-reversal permutation algorithm.

1. Decimation-in-time FFT
- Output is initialized as the input in the bit-reversal ordering
- Cooley-Tukey butterfly is performed on the output

It leads to an O(N*logN) complexity.

Note, we consider particular cases where the input is padded by zeros that always happens in erasure encoding process. Normally, FFT process starts performing butterfly operations on groups of size 2, then 4, 8 etc. Thanks to zero padding, we can start from a group size where there is at most a non-zero element of input in the group, all other elements of the group are zero. Indeed, after performing butterfly operation on the group, all elements equal to the non-zero one.

Given FFT length N, number of data length K, the output is initialized in the following clever way:
- Each group contains N/K elements
- Perform on groups containing a non-zero element of input: copy input's element at the bit-reversed index to all elements of the correspondent group of output
- For other groups, initialize them normally

It leads to an O(N*logK) complexity

2. Decimation-in-frequency FFT or inverse FFT
- Gentleman-Sande butterfly is performed on the input
- Output is copied from input at bit-reversed indices

It leads to an O(N*logN) complexity.

### CODE REVIEW

File: src/gf_ring.h [hunk 1]
```diff
@@ -459,6 +461,61 @@ inline void RingModN<T>::sub_vecp_to_vecp(
...
+ * \f{eqnarray*}{
+ *  P_i &= P_i + c \times Q_i \\
+ *  Q_i &= P_i - c \times Q_i \\
+ * \f}
+ *
+ * @param coef - coefficient, a finite field element
+ * @param buf1 - a buffer of `len` elements
+ * @param buf2 - a buffer of `len` elements
+ * @param len - number of elements per buffer
+ * @return
```
Comment (slaperche-scality) [inline src/gf_ring.h]: if it returns nothing, then remove the `@return`

File: src/gf_ring.h [hunk 2]
```diff
@@ -459,6 +461,61 @@ inline void RingModN<T>::sub_vecp_to_vecp(
     }
 }
 
+/** Butterfly computation for Cooley-Tukey FFT algorithm
+ *
+ * Perform in-place oprations on two buffers `P`, `Q` with a coefficient `c`
+ * \f{eqnarray*}{
```
Comment (slaperche-scality) [inline src/gf_ring.h]: I didn't check the DOxygen output, but don't we need to skip a line before the `\f{eqnarray*}{`?

File: src/gf_ring.h [hunk 3]
```diff
@@ -459,6 +461,61 @@ inline void RingModN<T>::sub_vecp_to_vecp(
...
+ *  P_i &= P_i + Q_i \\
+ *  Q_i &= c \times (P_i - Q_i)
+ * \f}
+ *
+ * @param coef - coefficient, a finite field element
+ * @param buf1 - a buffer of `len` elements
+ * @param buf2 - a buffer of `len` elements
+ * @param len - number of elements per buffer
+ * @return
```
Comment (slaperche-scality) [inline src/gf_ring.h]: ditto

File: src/gf_ring.h [hunk 4]
```diff
@@ -459,6 +461,61 @@ inline void RingModN<T>::sub_vecp_to_vecp(
...
+        T b = mul(coef, buf2[i]);
+        buf1[i] = add(a, b);
+        buf2[i] = sub(a, b);
+    }
+}
+
+/** Butterfly computation for Gentleman-Sande FFT algorithm
+ *
+ * Perform in-place oprations on two buffers `P`, `Q` with a coefficient `c`
+ * \f{eqnarray*}{
```
Comment (slaperche-scality) [inline src/gf_ring.h]: ditto

File: src/gf_ring.cpp [hunk 5]
```diff
@@ -58,6 +58,108 @@ void RingModN<uint32_t>::mul_coef_to_buf(
...
+        for (size_t i = vec_len * ratio; i < len; ++i) {
+            uint16_t a = buf1[i];
+            uint16_t b = mul(coef, buf2[i]);
+            buf1[i] = add(a, b);
+            buf2[i] = sub(a, b);
+        }
+    }
+}
+
+template <>
```
Comment (slaperche-scality) [inline src/gf_ring.cpp]: This is a copy paste of the 16-bit version…

I wonder if we could factorize this kind of code using a type trait (`is_vectorizable`) + `std::enable_if`

File: src/gf_ring.cpp [hunk 6]
```diff
@@ -58,6 +58,108 @@ void RingModN<uint32_t>::mul_coef_to_buf(
...
+            uint16_t b = buf2[i];
+            uint16_t c = sub(a, b);
+            buf1[i] = add(a, b);
+            buf2[i] = mul(coef, c);
+        }
+    }
+}
+
+template <>
+void RingModN<uint32_t>::butterfly_gs(
```
Comment (slaperche-scality) [inline src/gf_ring.cpp]: ditto

File: src/gf_ring.cpp (hunk 5 above)
Comment (slaperche-scality) [reply src/gf_ring.cpp]: I'll handle it in #222

File: src/fft_2n.h [hunk 7]
```diff
@@ -39,44 +40,43 @@
 #include "vec_vector.h"
 #include "vec_zero_ext.h"
 
+/** Compute bit-reversed number for a given number
```
[the remark this answers is not in the document]
Comment (lamphamsy) [reply src/fft_2n.h]: we do agree that this change is not necessary as the gain is not significant and this computation is performed once

**Comment by lamphamsy** [comment]
@slaperche-scality : thanks for your review. It's ready for next one :)

**Comment by lamphamsy** [comment]
Merging the PR as all comments are addressed.

**Comment by lamphamsy** [comment]
Oh, my bad. I forgot to change the base branch.

**Pull request merged**